### ISSUE TITLE
[Help Wanted] Running MCMC under the prior also removes the node calibrations

### ISSUE THREAD

**Issue by bjoelle** [revbayes/revbayes#463]
Running the `mcmc.run` with `underPrior = TRUE` seems to remove all clamped nodes, including node calibrations (at least if node calibrations are set following this tutorial [here](https://revbayes.github.io/tutorials/relative_time_constraints/))
I don't think this is proper behaviour as node calibrations are generally considered part of the prior - in particular this makes difficult to check the effective priors on the calibrations, which is generally good practice

**Comment by bjoelle** [comment]
thanks, that is helpful to know !

**Comment by bredelings** [comment]
Yeah, "remove the dnPhyloCTMC and run the MCMC as usual" seems like it corresponds more closely to what BEAST people mean by "run under the prior".  What they really mean is "run without the sequence data", which is exactly what happens when you remove `dnPhyloCTMC( )`.

**Comment by davidcerny** [comment]
@bjoelle Do you think this issue can be closed now, or would you suggest adding a brief note about this behavior to the documentation for `mcmc.run`?

**Comment by bjoelle** [comment]
@davidcerny I don't think the core issue is resolved ? I don't think this is correct behaviour for this parameter, and I don't think documenting the bug is enough to make sure people will not run into it again...

**Comment by bredelings** [comment]
Hmm... I think the current behavior is correct.  The prior does not include calibration likelihoods.  BEAST people have a habit of calling the prior + calibration likelihoods "the prior", but it isn't a prior.  This is extra confusing because people like Sandra Álvarez-Carretero, Mario dos Reis, Bruce Rannala, etc. use a different framework where they modify the prior instead of adding calibration likelihoods.

It seems that the problem with this parameter is not whether it is technically correct (I think it is correct) but a matter of perception:  what do people expect it to do?  Right now it is confusing and misleading, given how (some) people use the phrase "under the prior".

**Comment by bjoelle** [comment]
I don't think this is related to BEAST culture to be honest - in my understanding, node calibrations are considered part of the prior because they're set by the user based on information which is external to the inference, which is pretty much the definition of a prior
certainly if you want to know the influence of model choices on the results of the inference you need to include them

**Comment by bredelings** [comment]
> certainly if you want to know the influence of model choices on the results of the inference you need to include them

True enough.  I think its more helpful to say "prior to what".  Let's consider an analysis with a birth-death prior, fossil information, and DNA data.  In this case using you can talk a distribution that is
1. prior to fossil information and DNA data
2. posterior to fossil information and prior to DNA data
3. posterior to fossil information and also posterior to DNA data

What you are describing is what I would call "posterior to fossil information and prior to DNA data" (#2).  Leaving out the fossil information is also a prior though.  That is what revbayes does at the moment when you set `underPRIOR = TRUE`.

I don't think putting two probability distributions on the same node makes sense mathematically.  I agree that you can run MCMC under a variable that has two PDF terms and that it can make sense, but I do not think it is possible to mathematically interpret both PDF terms as priors.

**Comment by bredelings** [comment]
So, one option would be to remove the `underPrior` argument.  The benefit is that nobody will misinterpret this as including node calibration likelihoods. The downside is that people who are not doing fossil analyses cannot use it to explore their prior.  And people doing fossil analyses cannot use it to explore the prior without node calibrations.

Another option would be to rename it something like `likelihoodOnly` that is less likely to be interpreted as meaning "run under the prior + fossil information".  Any good name proposals?

A third option would be to add a heat parameter `beta` that raises the likelihood to the power `beta`, as in a heated chain.  Setting `beta = 0` would make the likelihood go away.  Setting `beta=1` would result in the normal posterior.

**Comment by davidcerny** [comment]
I tend to agree with Ben – it's a common criticism of node-dating that calibration densities are basically pseudodata rather than priors in the usual sense of the word, because there's no information in the sequence data that could possibly update them (e.g. Heath & Moore 2014). The fact we actually do treat them as data makes that property explicit, even thought it might sound strange to people who are used to other software.

More importantly, even if we agreed this is a bug, would it even be fixable? RevBayes would have to have some way of knowing which clamped nodes belong to calibrations, so that it can keep them clamped while unclamping all others. I'm not sure that information can be gleaned from the model DAG. Alternatively, we could change the behavior of `underPrior=TRUE` so that it drops any stochastic nodes that were drawn from `dnPhyloCTMC`. But that would lead to unexpected (and IMO, incorrect) behavior in analyses that don't involve any character data to begin with.

@bredelings Re your last comment: how would that be different from `likelihoodHeat`, which is already an existing argument of the `mcmc` function?

**Comment by bredelings** [comment]
@davidcerny Oh, I didn't realize we already had a heat function :-P I guess the only difference is that one is an argument to `mcmc` and one is an argument to `mcmc.run`?

**Comment by davidcerny** [comment]
P.S. Forgot to say it, but since the fundamental issue is precisely that calibrations are part of the likelihood and not the prior, I don't think renaming `underPrior` to `noLikelihood` (I assume that's what you meant?) or raising the likelihood term to a power of 0 would address it.

**Comment by bjoelle** [comment]
@bredelings @davidcerny I see your point about the pseudodata, but for me the fact that sequence data cannot override the calibrations is exactly the reason they should be treated as part of the model choices (just like e.g. the sequence data cannot override your choice of a strict clock) 
but either way this feels like a very academical discussion, when in practical terms the reality is that people do need to check what the influence of their calibrations is (regardless of how we want to call that :p), and at the moment it doesn't seem easily possible in the RevBayes structure ?

maybe this can be addressed explicitly by the tutorials instead ? on the RB side I would support the renaming + possibly adding a print message at the beginning which lists the elements that were removed from the calculation just so it's clear what happens

**Comment by bredelings** [comment]
> but either way this feels like a very academical discussion, when in practical terms the reality is that people do need to check what the influence of their calibrations is (regardless of how we want to call that :p), and at the moment it doesn't seem easily possible in the RevBayes structure ?

Right now, I guess what people have to do is to comment out all the phyloctmc terms.  So, where would we put that in the tutorials?

> @bredelings @davidcerny I see your point about the pseudodata, but for me the fact that sequence data cannot override the calibrations is exactly the reason they should be treated as part of the model choices (just like e.g. the sequence data cannot override your choice of a strict clock)

I guess the issue here is that the sequence data only provides information on rate*time, whereas the fossil data provides information on time directly.  So, you cannot increase precision on times simply by adding more sequence data.

**Comment by davidcerny** [comment]
How about this: keep the current name and current behavior for `underPrior = TRUE`, but make it print out a list of the nodes it removes / unclamps like @bjoelle suggests. Then, add a new option, `suppressCharacterData`, which removes all nodes that are of type `AbstractHomologousDiscreteCharacterData` from the model graph.

My sense is that the former would be pretty simple, but the latter may be harder. Then again, I've been able to avoid messing with RevBayes' DAG-parsing machinery so far, so I probably don't have a good sense of what it would take...

> So, where would we put that in the tutorials?

Good question – our current node-dating tutorials are a bit of a mess, and maybe there are other problems to be fixed first. E.g., [this tutorial](https://revbayes.github.io/tutorials/dating/nodedate) (which contains a section called "Running the analysis under the prior" that we might want to rewrite) clamps calibrations to negative values, @mlandis's tutorial [here](https://revbayes.github.io/tutorials/timefig_dating/#biogeographic-dating-with-node-calibrations) clamps them to 0, and [this tutorial](https://github.com/revbayes/revbayes_tutorial/blob/master/tutorial_TeX/RB_DivergenceTime_Calibration_Tutorial/RB_DivergenceTime_Calibration_Tutorial_content.tex) which Michael references and which is no longer up on the website clamps them to positive values.

**Comment by bjoelle** [comment]
and the troubleshooting tutorial [here](url) (which is what I was writing when I filed the issue) uses `setValue` instead of `clamp` for the calibrations because that seemed like the least difficult to explain
we should probably agree on **one** method at some point hum

**Comment by bredelings** [comment]
> Then, add a new option, suppressCharacterData, which removes all nodes that are of type AbstractHomologousDiscreteCharacterData from the model graph.

Will that be sufficient?  Could be other nodes that people might want to suppress if the goal is to know how the fossil calibrations interact?

> and the troubleshooting tutorial [here](https://github.com/revbayes/revbayes/issues/url) (which is what 
> I was writing when I filed the issue) uses setValue instead of clamp for the calibrations because that
> seemed like the least difficult to explain
> we should probably agree on one method at some point hum
Yeah.  This is a bit orthogonal to the current issue, but I'm going to submit a Pr to implement the pseudodata approach:
```
fossilInfo ~ pseudoData(node_time)
fossilInfo.clamp( dnExponential(rate) |> tnShift(offset))
```
So perhaps this could be the proposed single approach.

(As a side note to the side note, it looks like `dnExponential` has an `offset` parameter, but it doesn't show up in the help.)

Mathematically, we should NOT be normalizing these pseudo data distributions to sum to one.  But that won't cause trouble unless they have parameters that are also being estimated.

**Comment by davidcerny** [comment]
> Will that be sufficient? Could there be other nodes that people might want to suppress if the goal is to know how the fossil calibrations interact?

I'm not sure – what could those be? Usually, when calibrations are treated as part of the prior, the goal is to sample from the "effective prior" that's induced by interactions between (1) the calibrations and the birth-death prior and (2) the individual calibrations themselves. This is functionally equivalent to running the analysis without sequence / character data, as Sebastian pointed out. I guess there is a lot of other nodes that have to do with substitution models and stuff, but those should just end up being disconnected from the graph if we don't make that final draw from `dnPhyloCTMC`, no?

> Yeah. This is a bit orthogonal to the current issue, but I'm going to submit a Pr to implement the pseudodata approach:

Yeah, I think this is the most elegant approach of all.

**Comment by davidcerny** [comment]
This is now solved by PR #628. We just need to change the [node-dating tutorial](https://revbayes.github.io/tutorials/dating/nodedate) to make use of the new `.ignoreData` option, and to get rid of the "under the prior" terminology that @bredelings dislikes.
